cpp/src/arrow/filesystem/gcsfs.cc

```diff
@@ -200,6 +201,16 @@ class GcsOutputStream : public arrow::io::OutputStream {
     return internal::ToArrowStatus(stream_.last_status());
   }
 
+  Future<> CloseAsync() override {
+    if (closed_) return Status::OK();
+
+    auto self = std::dynamic_pointer_cast<GcsOutputStream>(shared_from_this());
```

Comment:
Can't you do this without the cast here? In the body of the lambda you can cast the `self.get()` pointer to `GcsOutputStream` to de-virtualize the call.

Reply:
```
class ARROW_EXPORT OutputStream : virtual public FileInterface, public Writable {
 protected:
  OutputStream() = default;
};
```

The `shared_from_this` is in `FileInterface`. I think we cannot.

Reply:
Can it be a `static_pointer_cast`, then?